pkg/gatewayserver/io/basicstationlns/correlations/correlations.go

```diff
@@ -0,0 +1,92 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package correlations
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// DownlinkInfo is the information associated with a particular downlink.
+type DownlinkInfo struct {
+	correlationIDs []string
+	txTime         time.Time
+}
+
+// DownlinkCorrelation is used to maintain tokens to correlate Downlink messages with their corresponding TxConfitrmations.
+type DownlinkCorrelation struct {
+	// token is the unique token associated with each Downlink.
+	// It's passed to the BasicStation as the `diid` field and is returned as-is in the TxConfirmation if the downlink packet was put on air.
+	// This is a free-running counter that is allowed to overflow and is cleaned up periodically by the garbage collector.
+	token        int64
+	expiration   time.Duration
+	ctx          context.Context
+	correlations sync.Map
```

Comment:
I don't think sync.Map is the most appropriate data structure here. I think a simple slice would do the trick just fine.

Reply:
We could do something like below. No need to work with an atomic counter and sync map. No need for a GC goroutine. We can set an upper bound on memory (here 64 correlations).

```go
type correlation struct {
	token int64
	ids   []string
}

type correlations struct {
	mu           sync.Mutex
	correlations []correlation
}

const maxSize = 64

func (c *correlations) Add(ids ...string) int64 {
	correlation := correlation{ids: ids}
	c.mu.Lock()
	if len(c.correlations) > 0 {
		correlation.token = c.correlations[len(c.correlations)-1].token + 1
	}
	c.correlations = append(c.correlations, correlation)
	if overflow := len(c.correlations) - maxSize; overflow > 0 {
		c.correlations = c.correlations[overflow:]
	}
	c.mu.Unlock()
	return correlation.token
}

func (c *correlations) Get(token int64) []string {
	c.mu.Lock()
	defer c.mu.Unlock()
	if len(c.correlations) == 0 {
		return nil
	}
	i := token - c.correlations[0].token
	if i < 0 || int(i) > len(c.correlations)-1 {
		return nil
	}
	correlation := c.correlations[i]
	return correlation.ids
}
```

Reply:
This implementation is similar to `io/udp` tokens but I do get your point.

We could set a smaller upper limit (Ex: 8 bits) and allow rollover since we won't have that many concurrent downlinks handled by one instance.

@johanstokking: thoughts?

Reply:
This is not similar to `io/udp` right? I mean that's a simple map with a gc and a rolling `uint32`.

Reply:
Yeah I agree we need a slice here; this has to become simpler, and I agree it would be simple and concise with a slice. When I implemented the UDP frontend I didn't realize that the map does not free up space.

Reply:
I think we should leave this correlation stuff out until we can use @johanstokking's implementation https://github.com/TheThingsNetwork/lorawan-stack/compare/feature/180-gs-schedule-abs-time#diff-132310a01664f102b53af8cf5aa08ee0